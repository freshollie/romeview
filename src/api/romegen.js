import axios from "axios";

axios.defaults.baseURL = "http://localhost:4943";

/**
 * Request the roman numerals
 * for the given decimal number.
 *
 * Throws any server errors
 *
 * @param {number | string} decimal
 */
export async function toNumerals(decimal) {
  return (await axios.get(`/decimal/${decimal}`)).data.numerals;
}

/**
 * Request the decimal form of the given roman
 * numerals.
 *
 * Throws any server errors
 *
 * @param {string} numerals
 */
export async function toDecimal(numerals) {
  return (await axios.get(`/numerals/${numerals}`)).data.decimal;
}
