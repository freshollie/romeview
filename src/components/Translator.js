import React, { Component } from "react";
import styles from "./Translator.module.css";

import * as romegen from "../api/romegen";

export default class Translator extends Component {
  // String constants
  OUTPUT_PLACEHOLDER = "Output";

  NUMERALS_TEXT = "Numerals";
  DECIMAL_TEXT = "Decimal";

  constructor(props) {
    super(props);

    this.state = {
      input: "",
      output: "",
      forwards: true,
      errorMessage: "",
      switchActive: false
    };

    this.handleInput.bind(this);
    this.switchDirection.bind(this);
  }

  /**
   * Switch the translation direction
   */
  switchDirection = () => {
    this.setState(
      { switchActive: true, forwards: !this.state.forwards },
      () => {
        this.handleInput(this.state.output);
      }
    );
  };

  showOutput(output) {
    this.setState({ output });
  }

  showErrorMessage(errorMessage) {
    this.setState({ errorMessage });
  }

  resetErrorMessage() {
    this.setState({ errorMessage: "" });
  }

  /**
   * Handle automatically updating
   * the output when we get any input
   */
  handleInput = async inputText => {
    this.setState({
      input: inputText
    });

    if (inputText.length === 0) {
      this.resetErrorMessage();
      this.showOutput("");
      return;
    }

    try {
      if (this.state.forwards) {
        this.showOutput(await romegen.toNumerals(inputText));
      } else {
        this.showOutput(await romegen.toDecimal(inputText));
      }
      this.resetErrorMessage();
    } catch (e) {
      // Pull errors from api
      if (e.response) {
        if (
          e.response.status === 422 &&
          e.response.data &&
          e.response.data.error
        ) {
          this.showErrorMessage(e.response.data.error);
        } else {
          this.showErrorMessage("Server error");
        }
      } else {
        this.showErrorMessage("Unknown error");
      }
    }
  };

  render() {
    return (
      <div className={styles.Translator}>
        <div className={styles.container}>
          <div className={styles["container-title"]} id="input-title">
            {this.state.forwards ? this.DECIMAL_TEXT : this.NUMERALS_TEXT}
          </div>
          <input
            className={
              styles.input +
              " " +
              (this.state.errorMessage.length > 0 ? styles["input-error"] : "")
            }
            type="text"
            value={this.state.input}
            onChange={e => this.handleInput(e.target.value)}
            placeholder={
              this.state.forwards
                ? "Type a decimal up to 3999"
                : "Type a set of Roman Numerals"
            }
          />
          <div className={styles.error}>{this.state.errorMessage}</div>
        </div>

        <button
          className={
            (this.state.switchActive ? [styles["swap-button-active"]] : "") +
            " " +
            styles["swap-button"]
          }
          onTransitionEnd={e => {
            this.setState({ switchActive: false });
          }}
          onClick={this.switchDirection}
        >
          {"<-->"}
        </button>

        <div className={styles.container}>
          <div className={styles["container-title"]} id="output-title">
            {this.state.forwards ? this.NUMERALS_TEXT : this.DECIMAL_TEXT}
          </div>
          <div className={styles["output-container"]}>
            <div className={styles.output}>{this.state.output}</div>
          </div>
        </div>
      </div>
    );
  }
}
