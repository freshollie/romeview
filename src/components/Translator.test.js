import React from "react";
import ReactDOM from "react-dom";
import flushPromises from "flush-promises";
import { shallow } from "enzyme";
import Translator from "./Translator";
import renderer from "react-test-renderer";

import * as romegen from "../api/romegen";

jest.mock("../api/romegen");

it("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<Translator />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it("renders correctly", () => {
  const tree = renderer.create(<Translator />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("has an input field", () => {
  const wrapper = shallow(<Translator />);
  expect(wrapper.find("input").exists()).toBeTruthy();
});

it("converts input to numerals on input", async () => {
  romegen.toNumerals.mockImplementation(async () => "MMII");

  const wrapper = shallow(<Translator />);
  wrapper.find("input").simulate("change", { target: { value: "2002" } });

  await flushPromises();

  expect(romegen.toNumerals).toHaveBeenCalledWith("2002");
  expect(wrapper.find(".output").text()).toBe("MMII");
});

it("allows input and output to be switched", async () => {
  const wrapper = shallow(<Translator />);

  expect(wrapper.find("#input-title").text()).toBe("Decimal");
  expect(wrapper.find("#output-title").text()).toBe("Numerals");

  wrapper.find("button").simulate("click");
  expect(wrapper.find("#input-title").text()).toBe("Numerals");
  expect(wrapper.find("#output-title").text()).toBe("Decimal");
});

it("allows numerals to be translated to decimals", async () => {
  romegen.toDecimal.mockImplementation(async () => 3922);

  const wrapper = shallow(<Translator />);

  // reverse the translator
  wrapper.find("button").simulate("click");
  wrapper.find("input").simulate("change", { target: { value: "MMMCMXXII" } });

  await flushPromises();

  expect(romegen.toDecimal).toHaveBeenCalledWith("MMMCMXXII");
  expect(wrapper.find(".output").text()).toBe("3922");
});

it("shows nothing in output when nothing entered", async () => {
  const wrapper = shallow(<Translator />);
  romegen.toNumerals.mockImplementation(async () => "MMIX");

  wrapper.find("input").simulate("change", { target: { value: "2009" } });

  await flushPromises();

  wrapper.find("input").simulate("change", { target: { value: "" } });
  await flushPromises();
  expect(wrapper.find(".output").text()).toBe("");
});

async function testErrorMessage(error, message) {
  romegen.toNumerals.mockImplementation(async () => {
    throw error;
  });
  const wrapper = shallow(<Translator />);

  wrapper.find("input").simulate("change", { target: { value: "MMMCMXXII" } });

  await flushPromises();

  expect(wrapper.find(".error").text()).toBe(message);
}

it("shows any errors from api", async () => {
  await testErrorMessage(
    {
      response: {
        status: 422,
        data: {
          error: "Some error"
        }
      }
    },
    "Some error"
  );
});

it("highlights input box on error", async () => {
  romegen.toNumerals.mockImplementation(async () => {
    throw new Error("Oops");
  });
  const wrapper = shallow(<Translator />);

  wrapper.find("input").simulate("change", { target: { value: "MMMCMXXII" } });

  await flushPromises();

  expect(wrapper.find("input").hasClass("input-error")).toBeTruthy();
});

it("shows error messages from failure", async () => {
  await testErrorMessage(
    {
      response: {
        status: 500
      }
    },
    "Server error"
  );

  await testErrorMessage(new Error("Idk what this is"), "Unknown error");
});

it("clears error messages when valid input given", async () => {
  romegen.toNumerals.mockImplementation(async () => {
    throw new Error("Some error");
  });
  const wrapper = shallow(<Translator />);
  wrapper.find("input").simulate("change", { target: { value: "MMMCMXXII" } });

  await flushPromises();

  expect(wrapper.find(".error").text()).toBe("Unknown error");

  romegen.toNumerals.mockImplementation(async () => {
    return "anything";
  });

  wrapper.find("input").simulate("change", { target: { value: "asasd" } });
  await flushPromises();

  expect(wrapper.find(".error").text()).toBe("");
});

it("swaps output to input when mode switched", async () => {
  romegen.toDecimal.mockImplementation(async () => 2001);
  romegen.toNumerals.mockImplementation(async () => "MMI");

  const wrapper = shallow(<Translator />);

  wrapper.find("input").simulate("change", { target: { value: "2001" } });
  await flushPromises();

  wrapper.find("button").simulate("click");
  await flushPromises();

  expect(wrapper.find("input").prop("value")).toBe("MMI");

  expect(romegen.toDecimal).toHaveBeenCalledWith("MMI");
  expect(wrapper.find(".output").text()).toBe("2001");
});

it("should set swap button to active when clicked", () => {
  const wrapper = shallow(<Translator />);

  wrapper.find("button").simulate("click");
  expect(wrapper.find("button").hasClass("swap-button-active")).toBeTruthy();
});

it("should set swap button to inactive once the transition finishes", () => {
  const wrapper = shallow(<Translator />);

  wrapper.find("button").simulate("click");
  wrapper.find("button").simulate("transitionEnd");
  expect(wrapper.find("button").hasClass("swap-button-active")).toBeFalsy();
});

it("always converts input to uppercase", () => {
  const wrapper = shallow(<Translator />);
  wrapper.find("input").simulate("change", { target: { value: "mi" } });

  expect(wrapper.find("input").prop("value")).toBe("MI");
});
