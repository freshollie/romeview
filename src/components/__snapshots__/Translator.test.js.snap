// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="Translator"
>
  <div
    className="container"
  >
    <div
      className="container-title"
      id="input-title"
    >
      Decimal
    </div>
    <input
      className="input "
      onChange={[Function]}
      placeholder="Type a decimal up to 3999"
      type="text"
      value=""
    />
    <div
      className="error"
    >
      
    </div>
  </div>
  <button
    className=" swap-button"
    onClick={[Function]}
    onTransitionEnd={[Function]}
  >
    &lt;--&gt;
  </button>
  <div
    className="container"
  >
    <div
      className="container-title"
      id="output-title"
    >
      Numerals
    </div>
    <div
      className="output-container"
    >
      <div
        className="output"
      >
        
      </div>
    </div>
  </div>
</div>
`;
